src/System.Net.Http/src/System/Net/Http/HttpMethod.cs

```diff
@@ -106,7 +122,12 @@ public override bool Equals(object obj)
 
         public override int GetHashCode()
         {
-            return _method.ToUpperInvariant().GetHashCode();
+            if (_hashcode == 0)
+            {
+                _hashcode = _method.ToUpperInvariant().GetHashCode();
```

Comment:
It's not actually 7 string comparisons. After a certain number of cases (>= 7 in Roslyn, which this falls under) the C# compiler will use a hash table (the VS2013 compiler actually uses a static cached `Dictionary<string, int>`, while Roslyn emits a more efficient approach).

I like @Eyas' suggestion. Though, `char.IsUpper()` is actually ~1.6x slower than the `switch` with Roslyn, and I'm a little concerned about potential esoteric edge cases (e.g. turkish i, etc.). But checking if each char is an uppercase ASCII character (which is what these methods should be per the HTTP RFC) is basically the same speed to slightly faster (~5%) as the `switch` for the known string cases, and significantly beats the switch for custom strings that are already ASCII uppercase: ~7.2x faster (x86) and ~6.5x faster (x64 RyuJIT).

Based on that, I'd be on board with getting rid of the private constructor, and instead just doing this in `GetHashCode()`:

```c#
public override int GetHashCode()
{
    if (_hashcode == 0)
    {
        // If _method is already uppercase, _method.GetHashCode() can be
        // used instead of _method.ToUpperInvariant().GetHashCode(),
        // avoiding the unnecessary extra string allocation.
        _hashcode = IsUpperAscii(_method) ?
            _method.GetHashCode() :
            _method.ToUpperInvariant().GetHashCode();
    }

    return _hashcode;
}

private static bool IsUpperAscii(string value)
{
    for (int i = 0; i < value.Length; i++)
    {
        char c = value[i];
        if (!(c >= 'A' && c <= 'Z'))
        {
            return false;
        }
    }

    return true;
}
```